fastchat/serve/bard_worker.py

```diff
@@ -0,0 +1,157 @@
+
+
+class Response(BaseModel):
+    content: str
+    factualityQueries: Optional[List]
+    textQuery: Optional[Union[str, List]]
+    choices: List[dict]
+    state: ConversationState
+
+
+class Chatbot:
+    """
+    A class to interact with Google Bard.
+    Parameters
+        session_id: str
+            The __Secure-1PSID cookie.
+    """
+
+    def __init__(self, session_id):
+        headers = {
+            "Host": "bard.google.com",
+            "X-Same-Domain": "1",
+            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36",
+            "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8",
+            "Origin": "https://bard.google.com",
+            "Referer": "https://bard.google.com/",
+        }
+        self.session = httpx.AsyncClient()
+        self.session.headers = headers
+        self.session.cookies.set("__Secure-1PSID", session_id)
+        self.SNlM0e = None
+
+    async def _get_snlm0e(self):
+        resp = await self.session.get(url="https://bard.google.com/", timeout=10)
+        # Find "SNlM0e":"<ID>"
+        if resp.status_code != 200:
+            raise Exception("Could not get Google Bard")
+        SNlM0e = re.search(r"SNlM0e\":\"(.*?)\"", resp.text).group(1)
+        return SNlM0e
+
+    async def ask(self, message: Message) -> Response:
+        """
+        Send a message to Google Bard and return the response.
+        :param message: The message to send to Google Bard.
+        :return: A dict containing the response from Google Bard.
+        """
+        if message.state.conversation_id == "":
+            message.state.req_id = int("".join(random.choices(string.digits, k=4)))
+        # url params
+        params = {
+            # "bl": "boq_assistant-bard-web-server_20230315.04_p2",
+            # This is a newer API version
+            "bl": "boq_assistant-bard-web-server_20230507.20_p2",
+            "_reqid": str(message.state.req_id),
+            "rt": "c",
+        }
+
+        # message arr -> data["f.req"]. Message is double json stringified
+        message_struct = [
+            [message.content],
+            None,
+            [
+                message.state.conversation_id,
+                message.state.response_id,
+                message.state.choice_id,
+            ],
+        ]
+        data = {
+            "f.req": json.dumps([None, json.dumps(message_struct)]),
+            "at": self.SNlM0e,
+        }
+
+        # do the request!
+        resp = await self.session.post(
+            "https://bard.google.com/_/BardChatUi/data/assistant.lamda.BardFrontendService/StreamGenerate",
+            params=params,
+            data=data,
+            timeout=60,
+        )
+
+        chat_data = json.loads(resp.content.splitlines()[3])[0][2]
+        if not chat_data:
+            return Response(
+                content=f"Google Bard encountered an error: {resp.content}.",
```

Comment:
Just to confirm: if Bard fails, is user still allowed to vote? if so, we should ban it right?

Reply:
we will figure such votes later